src/_base/harness/config/cleanup.yml

```
@@ -0,0 +1,20 @@
+function('built_images', [services]): |
+  #!php
+  $builtImages = [];
+
+  foreach ($services as $service) {
+    if ($service['image'] && count($service['upstream']) > 0) {
+      $builtImages[] = $service['image'];
+    }
+  }
+
+  # workspace commands don't allow non-string types
+  = join(' ', $builtImages);
+
+command('cleanup built-images'):
+  env:
+    IMAGES: = built_images(docker_service_images())
+  exec: |
+    #!bash
+    IMAGES=($IMAGES)
+    docker image rm --force -- "${IMAGES[@]}"
```

This seems to catch the stragglers without needing to prune, though it deletes nginx and php-fpm early and then complains they don't exist later:
```suggestion
    for image in ${IMAGES[@]}; do
      docker images --filter=since="${image}" -q | xargs --no-run-if-empty docker image rm --force
    done
    docker image rm --force -- "${IMAGES[@]}"
```

This is currently running in dynamic mode too so need to drop `--no-run-if-empty` if on macOS at the risk of an error occurring if no entries:

```suggestion
    for image in ${IMAGES[@]}; do
      docker images --filter=since="${image}" -q | xargs docker image rm --force
    done
    docker image rm --force -- "${IMAGES[@]}"
```

---

I think a prune is fine, it only deletes dangling images anyway

---

only the LABEL way tho would be suitable for concurrent executors, both filter=since and prune would have issues

---

I've added the LABEL way as is good for local too, but I think if there's good enough reason for anything more then for CI it be prune